### PR TITLE
Make `unconditional_recursion` warning detect recursive drops

### LINKED ISSUE #50049: Detect unconditional recursion in Drop
If you instantiate the same struct in the Drop function, the problem will loop indefinitely and crash.

The Code

`
struct ToDrop;

impl Drop for ToDrop {
    fn drop(&mut self) {
        let m = ToDrop;
        println!( "ToDrop is being dropped" );
    }
}
fn main() {
    let x =  ToDrop;
}
`

Well, I expected the compiler to catch this at compilation. So this could be an area of improvement.

Here is another variation that also causes a stackoverflow without intializing the same variable

`
struct A;
struct B;

impl Drop for A {
    fn drop( &mut self ) {
        let b = B;
    }
}

impl Drop for B {
    fn drop( &mut self ) {
        let a = A;
    }
}

fn main() {
    let a = A;
}
`

The error message at execution

`
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
Abort trap: 6
`

### LINKED ISSUE #55388: Should there be a warning at compile time when implementing Drop trait with creation of same type inside?
I'm fairly new to Rust and have been working through the reference book recently, I've just covered the Drop trait implementation and thought it would be fun to see what happens if I create a new instance of the type I just dropped inside of the function. 

Obviously this is completely nonsensical as nobody is going to do this but as you can probably tell this caused my stack to overflow and crash the program, would an error message at compile time be appropriate for something like this? 

I can't see why anyone would want to initialise new variables inside of the drop trait but it's possible.

Example code of my implementation:
```
struct CustomSmartPointer {
    data: String,
}

impl CustomSmartPointer{
    fn new(s: String) -> CustomSmartPointer {
        CustomSmartPointer{
            data: s,
        }
    }
}

impl Drop for CustomSmartPointer {
    fn drop(&mut self){
        println!("Dropping & creating new pointer");
        let c = CustomSmartPointer::new(String::from("test"));
        println!("New pointer created in memory");
    }
}

fn main() {
    let c = CustomSmartPointer::new(String::from("test"));
}
```

### PR DESCRIPTION
**Pull request by Enselic**
Closes #55388

Also closes #50049 unless we want to keep it for the second example which this PR does not solve, but I think it is better to track that work in #57965.

r? @oli-obk since you are the mentor for #55388

Unresolved questions:
- [x] There are two false positives that must be fixed before merging (see diff). I suspect the best way to solve them is to perform analysis after drop elaboration instead of before, as now, but I have not explored that any further yet. Could that be an option? **Answer:** Yes, that solved the problem.

@rustbot label +T-compiler +C-enhancement +A-lint

### CODE REVIEW

**Comment by Enselic** [comment]
I'll explore ways to fix the false positives before I send this to review. (Guidance on how to do it still appreciated.)

@rustbot author

File: tests/run-make/unconditional-drop-recursion/main.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+//! This can not be a regular ui test since we can only lint drop recursion
+//! after the drop elaboration MIR pass, which is not performed in ui tests.
```
Comment (Urgau) [inline tests/run-make/unconditional-drop-recursion/main.rs:2]: Can't you use the `build-error` directive to achieve the same goal with UI tests ?
```suggestion
// build-error
```

Comment (Enselic) [reply tests/run-make/unconditional-drop-recursion/main.rs:2]: I think you're right, thanks! Will fix. (Assuming you mean [`build-fail`](https://rustc-dev-guide.rust-lang.org/tests/ui.html#controlling-passfail-expectations))

**Comment by Enselic** [comment]
Moving the check for drop recursion until after drop elaboration indeed fixed the false positives. So I consider this ready for review now.

We probably want to do a perf run? Probably best to wait until after at least an initial review round. Maybe we want to refactor the code to avoid `rustc_mir_transform` depending on `rustc_mir_build`? Just let me know.

@rustbot ready

**Comment by cjgillot** [comment]
`mir_drops_elaborated_and_const_checked` does not run on all MIR bodies, so moving the lint there will cause it not to run. In particular in `check` builds on non-generic non-inline functions.

**Comment by cjgillot** [comment]
That said, I don't see how else run this lint after drop-elaboration.

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
Nominating for lang team discussion. This extends the `unconditional_recursion` lint to also detect when a `Drop` impl causes values of the same type getting dropped unconditionally. The one caveat being, that this is only reported in full builds, not in check builds. The existing lint is unaffected.

**Comment by Enselic** [comment]
(Just so we all are on the same page: that phenomenon is old and not introduced by this PR. See https://github.com/rust-lang/rust/issues/49292)

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

Sounds good, starting FCP!

**Comment by pnkfelix** [comment]
(once T-lang finishes FCP, we can relabel this back to T-compiler again)

**Comment by scottmcm** [comment]
Personally, I'm happy for the lint to detect all cases of unconditional recursion, wherever or whenever, without extra lang confirmation.  (I see lang as saying "yes, unconditional recursion is worth linting about, and compiler reviewers as sufficient for expanding that to catch more places of the same thing.)

The check/build difference is a shame, but false negatives are normal for lints, so I don't see it as a big issue.  (Unlike for *error*s.)

@rfcbot reviewed

**Comment by oli-obk** [comment]
I checked niko's and josh's box as both of them tried to fcp merge this PR: https://github.com/rust-lang/rust/pull/113902#issuecomment-1650375284 and https://github.com/rust-lang/rust/pull/113902#issuecomment-1650375291

**Pull request merged**